app/create_standby/common.py

```diff
@@ -0,0 +1,364 @@
+        """Get a hardlink tracker from the register.
+
+        Args:
+            _identifier: a string that can identify a group of hardlink file.
+            path: path that the caller wants to save file to.
+            nlink: number of hard links in this hardlink group.
+
+        Returns:
+            A hardlink tracker and a bool to indicates whether the caller is the writer or not.
+        """
+        with self._lock:
+            _ref = self._hash_ref_dict.get(_identifier)
+            if _ref:
+                _tracker = self._ref_tracker_dict[_ref]
+                return _tracker, False
+            else:
+                _ref = _WeakRef()
+                _tracker = _HardlinkTracker(path, _ref, nlink - 1)
+
+                self._hash_ref_dict[_identifier] = _ref
+                self._ref_tracker_dict[_ref] = _tracker
+                return _tracker, True
+
+
+class RegularInfSet(OrderedDict[RegularInf, None]):
+    """Use RegularInf as key, and RegularInf use path: Path as hash key."""
+
+    def add(self, entry: RegularInf):
+        self[entry] = None
+
+    def remove(self, entry: RegularInf):
+        del self[entry]
+
+    def iter_entries(self) -> Iterator[Tuple[bool, RegularInf]]:
+        _len, _count = len(self), 0
+        for entry, _ in self.items():
+            _count += 1
+            yield _len == _count, entry
+
+
+class RegularDelta(Dict[str, RegularInfSet]):
+    def __init__(self):
+        super().__init__()
+        # for fast lookup regularinf entry
+        self._pathset: Set[Path] = set()
+
+    def __len__(self) -> int:
+        return sum([len(_set) for _, _set in self.items()])
+
+    def add_entry(self, entry: RegularInf):
+        self._pathset.add(entry.path)
+
+        _hash = entry.sha256hash
+        if _hash in self:
+            self[_hash].add(entry)
+        else:
+            _new_set = RegularInfSet()
+            _new_set.add(entry)
+            self[_hash] = _new_set
+
+    def merge_entryset(self, _hash: str, _other: RegularInfSet):
+        if _hash not in self:
+            return
+
+        self[_hash].update(_other)
+        for entry, _ in _other.items():
+            self._pathset.add(entry.path)
+
+    def contains_hash(self, _hash: str) -> bool:
+        return _hash in self
+
+    def contains_path(self, path: Path):
+        return path in self._pathset
+
+
+@dataclass
+class DeltaBundle:
+    """NOTE: all paths are canonical!"""
+
+    # delta
+    rm_delta: List[str]
+    new_delta: RegularDelta
+    new_dirs: OrderedDict[DirectoryInf, None]
+
+    # misc
+    ref_root: Path
+    recycle_folder: Path
+    total_regular_num: int
+
+
+class DeltaGenerator:
+    # entry under the following folders will be scanned
+    # no matter it is existed in new image or not
+    FULL_SCAN_PATHS = set(
+        [
+            "/lib",
+            "/var/lib",
+            "/usr",
+            "/opt/nvidia",
+            "/home/autoware/autoware.proj",
+        ]
+    )
```

Comment:
For `/var`, I think only `/var/lib` should be fully scanned as other folders will change when the system is running, like `/var/run` or `/var/cache`.
For `/home`, I am not sure what kind of changes/new files will be introduced when the system is running, too much/too large files will significantly slow down the scan process.
The same as `/opt`.

Typically, I would like to fully scan the folders that likely not to change too much during the system running. 
Otherwise a normal path based scan(only scan files that the paths present both in old and new image) will be used.

Reply:
Currently is OK, but I'd like to introduce the following function (pseudo code):
```python
# before calling this file, src and ensure dst size should be the same.
def copy_with_sha256(src, dst):
  hash = sha256()
  with open(src, "rb") as s:
    with open(dst, "wb") as d:
      buf = s.read(CHUNK)
      if len(buf) == 0:
        break
      hash.update(buf)
      d.write(buf)
  return hash.digest()
```
If we can introduce this, the problem `what kind of changes/new files will be introduced` disappeared and the problem is not only for the files under /home. Basically, all files under the active partition should be verified by this function.
So it means the current implementation(aka legacy mode) has potentially bug.